File: BinaryPlay/940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();

        unordered_map<int , int> basket ;

        int i = 0 , j = 0 ;
        int maxLen = 0;

        while(j < n){
            basket[fruits[j]]++; //take the fruit

            //if basket reached the limit greater than 2 then take off the fruit
            if(basket.size() > 2){
                basket[fruits[i]]--;
                if(basket[fruits[i]] == 0){
                    basket.erase(fruits[i]);
                }
                i++;
            }else{
                maxLen = max(maxLen , (j - i + 1));
            }
            j++;
        }
        return maxLen;

    }
};

File: BinaryPlay/940-fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int n = fruits.length;

        Map<Integer , Integer> basket = new HashMap<>();
        int i = 0 , j = 0 ;
        int maxLen = 0;

        while(j < n){
            //basket that can store fruit of type 2
            basket.put(fruits[j] , basket.getOrDefault(fruits[j] , 0) + 1);
            
            //shrink if basket store greater than 2
            if(basket.size() > 2){
                int left = basket.get(fruits[i]) - 1;
                if(left == 0){
                    basket.remove(fruits[i]);
                }else{
                    basket.put(fruits[i] , left);
                }
                i++;
            }else{
                maxLen = Math.max(maxLen , j - i + 1);
            }
            j++;
        }
        return maxLen;
    }
}